libbpf-tools/doublefree.c

```diff
@@ -0,0 +1,593 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2022 LG Electronics */
+
+// 19-Oct-2022   Bojun Seo   Created this.
+#include <argp.h>
+#include <stdio.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "doublefree.h"
+#include "doublefree.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+//#define WEBOS true
+
+enum log_level {
+	DEBUG,
+	INFO,
+	WARN,
+	ERROR,
+};
+
+void __p(enum log_level level, char *level_str, char *fmt, ...);
+void set_log_level(enum log_level level);
+
+#define p_debug(fmt, ...) __p(DEBUG, "Debug", fmt, ##__VA_ARGS__)
+#define p_info(fmt, ...) __p(INFO, "Info", fmt, ##__VA_ARGS__)
+#define p_warn(fmt, ...) __p(WARN, "Warn", fmt, ##__VA_ARGS__)
+#define p_err(fmt, ...) __p(ERROR, "Error", fmt, ##__VA_ARGS__)
+
+int get_libc_path(char *path);
+pid_t execute_process(char *cmd);
+
+#define ON_MEM_FAILURE(buf) \
+if (NULL == buf) { \
+	p_err("FATAL: Failed to allocate memory on %s", __func__); \
+	exit(-1); \
+}
+
+/* default log level */
+static enum log_level log_level = ERROR;
+
+void sig_handler(int signo)
+{
+	if (signo == SIGINT || signo == SIGTERM)
+		kill(0, SIGKILL);
+}
+
+void __p(enum log_level level, char *level_str, char *fmt, ...)
+{
+	va_list ap;
+
+	if (level < log_level)
+		return;
+	va_start(ap, fmt);
+	fprintf(stderr, "%s: ", level_str);
+	vfprintf(stderr, fmt, ap);
+	fprintf(stderr, "\n");
+	va_end(ap);
+	fflush(stderr);
+}
+
+void set_log_level(enum log_level level)
+{
+	log_level = level;
+}
+
+int get_libc_path(char *path)
+{
+	FILE *f = NULL;
+	char buf[PATH_MAX] = {0, };
+	char *file_name = NULL;
+	float version = 0.0f;
+
+	f = fopen("/proc/1/maps", "rt");
+	if (f == NULL) {
+		p_err("Failed to open /proc/1/maps");
+		return -1;
+	}
+
+	while (fscanf(f, "%*x-%*x %*s %*s %*s %*s %[^\n]\n", buf) != EOF) {
+		if (strchr(buf, '/') != buf)
+			continue;
+
+		file_name = strrchr(buf, '/') + 1;
+#ifdef WEBOS
+		if (sscanf(file_name, "libc-%f.so", &version) == 1) {
```

Comment:
This issue is happening because of the difference of the path among platforms.
I think it would be better to add `libc` option like `gethostlatency` tool.

Reply:
I remove the whole `libc_path` things rather use `get_pid_lib_path` written in `uprobe_helpers.c`.
It is nice portable function!